### DISCUSSION TITLE
Could the 'show context lines' toggle update the results list immediately?

### DISCUSSION THREAD

**Discussion by Zeego** [dnGrep/dnGrep#1135] [Ideas]
First of all, thanks for making this awesome software! I tried it before many years ago, but for some reason I stuck with other software instead. I've just come back to trying it again and am very impressed!

So my suggestion: what would be awesome is when toggling the 'show context lines' that it refreshes the results list (for all the files where the actual results are expanded) in real-time. Currently, this toggle only has any effect if you re-run the search, which is a bit cumbersome.

**Comment by doug24** [comment]
Here is the original request that added that feature #361 - previously you could only change these values in the Options dialog.  

The context lines are read and cached when a search result is expanded in the results view. dnGrep does not cache the whole file, only matched line and context lines. To make it dynamic, the results tree be emptied and recreated, collapsing all the nodes and re-reading the files as the user reopens the file nodes in the tree.  I'm not convinced this is much of an improvement over the existing behavior.  I don't see getting to a place where you can simply toggle showing the context lines on and off.

**Comment by Zeego** [comment]
It appears to me that the context lines are cached at the time of the search, not when a result is expanded.

So if I have 'show context lines' enabled, then do the search, then disable 'show context lines' - every search result I expand already shows the context lines.

If all those context lines are already cached, then I'm hoping it should be possible to redraw the results list with or without the context lines without too much difficulty when toggling the setting?

**Comment by doug24** [comment]
There are three properties: ShowLinesInContext (Boolean), ContextLinesBefore (number), and ContextLinesAfter (number).  These properties are stored in the dnGrep settings which is a singleton class used across multiple assemblies and saved/restored in a file.  The properties are updated from the UI to the settings at the start of each search and when the Options dialog is closed.  So, changing any of the three properties on the main viewer doesn't update the settings until after the start of the next search.

The lines results tree gets populated as you open the file node, using the last saved settings - so changing the values of these properties does not have any effect until they get pushed to the dnGrep settings.  But what if they were updated immediately?  It is not about just redrawing the results list - all those files are not stored in memory, you have to go back and reread the files.

Let's say that you change ShowLinesInContext from false to true.  Now all the file nodes you open after that will have any lines in context.  But what do we do with the file nodes you've already opened?  Should they be left as-is, or should they be modified to add the context lines - should we go back and reread each of those files to get the context lines?

How about changing the ContextLinesBefore from 2 to 5?  Now we need to go back and reread all the opened file nodes to get additional lines.  

If we are going to go back to reread the files, when should that happen?  You probably do not want to do it immediately after one of the three properties change, because the user could be changing more than one property, and you don't want to reread the files multiple times.  It seems that when any of the properties change, we would have to close all the open file nodes and discard the cached lines so we can reread the files only when the user opens the node again.

An alternate implementation would be to read and keep all the lines of all the opened file nodes in memory, then just mark lines as a match line, a context line or a hidden line.  But you have to consider that search results can have a large number of files and large files. That would consume lots of memory.  Usually, you are finding matches on only a small number of lines out of each file.

**Comment by Zeego** [comment]
I totally understand what you're saying - I'm a developer myself. Having played about with it a bit more, here's where I'm up to:

1. You're right in that the context lines are read when a file node is expanded (sorry, I was wrong in my previous comment).
2. I agree that refreshing the file list (for all expanded files) for each change of `ContextLinesBefore` or `ContextLinesAfter` could be quite expensive.
3. It's interesting that the values are re-read when the Options dialog is closed (which then returns you to the main screen almost instantly).

Number 3 is particularly interesting, because that behaves in a way like I was asking for initially (albeit with the cumbersome wait to open the Options dialog in the first place).

So here's my suggestion: **always** _read_ (and cache) the context lines when you expand a file node. But if `ShowLinesInContext` is false, don't actually _display_ them. Then, if a user toggles `ShowLinesInContext`, you 'only' need to redraw the search results by adding in the (already cached) context lines, i.e. you don't need to re-read any files.

As for the `ContextLinesBefore` or `ContextLinesAfter` values, maybe keep it the same as it is now, i.e. changing these values only has any effect when you re-run the search.

_I appreciate that these changes I'm suggesting may need a slight redesign of the UI of the main screen to make sense to a user._

**Comment by doug24** [comment]
> It's interesting that the values are re-read when the Options dialog is closed (which then returns you to the main screen almost instantly).

It is the settings values that are updated when the Options dialog is closed.  Not the content of the current files.  So, the behavior is what I described above, depending on when you expand a file in the results tree, some will have context lines, and some will not.

I can easily implement changing the global settings when you modify any of the three properties at the bottom of the main window.  That will affect nodes not yet opened and have no impact on nodes already opened.  It seemed inconsistent to me, so I didn't implement it that way originally.  But if that is helpful, I think it would be easy to do.

I had also thought about always reading and caching the lines before/lines after even if ShowLinesInContext is false.  I wonder if anyone is setting ShowLinesInContext to false as an efficiency to save processing and memory on large files, and we would be undoing that with no visibility.  They could still do it by setting lines before and lines after to 0.  Or maybe I'm just overthinking it.

I'll take a look at doing this.

**Reply by Zeego** [reply]
I don't suppose there has been any further thought on this...?

**Reply by doug24** [reply]
It was added in [v4.2.46](https://github.com/dnGrep/dnGrep/releases/tag/v4.2.46.0).

**Reply by Zeego** [reply]
Ha! Just goes to show I should've checked for 'new' releases first! Excellent, thanks, I'll give it a try
